reason_4.0/lib/core/minisite_templates/modules/events.php

```diff
@@ -3940,6 +3946,93 @@ function get_item_ical_link($e, $all_ocurrences = true)
+		$title = $this->string_to_external_url_safe($e->get_value('name'));
+		//Create location string:
+		$location = $this->string_to_external_url_safe($e->get_value('location'));
+		$address = $this->string_to_external_url_safe($e->get_value('address'));
+		$location .= (!empty($location) && !empty($address)) ? ', ' : '';
+		$location .= (!empty($address)) ? $address : '';
+		//Create details string:
+		$content = $this->string_to_external_url_safe($e->get_value('content'));
+		$info_url = $this->string_to_external_url_safe('For more information, visit: '.$e->get_value('url'));
+		$cal_url = $this->string_to_external_url_safe('Event source: '.REASON_HOST.$this->parent->pages->get_full_url($this->page_id).'?event_id='.$e->id());
+		$details = (!empty($content)) ? $content.'%0A%0A' : '';
+		$details .= (!empty($e->get_value('url'))) ? $info_url.'%0A%0A'.$cal_url : $cal_url;
+		switch ($calendar) {
+			case 'google':
+				//Complete additional formating:
+				$dateTime->add(new DateInterval('PT'.$e->get_value('hours').'H'.$e->get_value('minutes').'M'));
+				$endDate = $dateTime->format('Ymd');
+				$endTime = $dateTime->format('His');
+				$dates = $all_day ? $startDate.'/'.$endDate : $startDate.'T'.$startTime.'Z/'.$endDate.'T'.$endTime.'Z';
+				//Create string to return:
+				$ret = 'http://www.google.com/calendar/event?action=TEMPLATE';
+				$ret .= '&dates='.$dates.'&ctz='.REASON_DEFAULT_TIMEZONE.'&text='.$title;
+				$ret .= (!empty($location)) ? '&location='.$location : '';
+				$ret .= (!empty($details)) ? '&details='.$details : '';
+				break;
+			case 'yahoo':
+				//Complete additional formating:
+				$duration = '';
+				if (!$all_day) {
+					$duration = ($hours < 10) ? '0'.$hours : $hours;
+					$duration .= ($minutes < 10) ? '0'.$minutes : $minutes;
+				}
+				$startDateTime = $all_day ? $startDate : $startDate.'T'.$startTime.'Z';
+				//Create string to return:
+				$ret = 'http://calendar.yahoo.com/?v=60&ST='.$startDateTime.'&TITLE='.$title;
+				$ret .= (!empty($duration)) ? '&DUR='.$duration : '';
+				$ret .= (!empty($location)) ? '&in_loc='.$location : '';
+				$ret .= (!empty($details)) ? '&DESC='.$details : '';
+				break;
+			default:
+				trigger_error('External calendar of "'.$calendar.'" given is not implemented. Unable to return url.');
+				$ret = NULL;
+		}
+		return $ret;
+	}
+	/**
+	 * Returns given string converted for use as a parameter value in urls.
+	 * Operations: Format lists, strip tags, straight to curly quotes, replace ampersand, urlencode, other minor changes.
+	 *
+	 * @param string $e string to convert
+	 * @return string converted string
+	 */
+	function string_to_external_url_safe($string) {
+		$string = preg_replace('/<li[^>]*>/',' • ',$string); //not sure how ol or levels, so simple ul bullets
+		$string = strip_tags($string);
+		$string = preg_replace('/(.*?)"(.*?)"(.*?)/','$1“$2”$3',$string); //curly quotes
+		$string = str_replace(array('"','&amp;'), array('“','and'), $string);
```

Comment:
This would be problematic for non-english text, e.g. "Huevos & patatas" becomes "Huevos and patatas" instead of the more appropriate "Huevos y patatas". Overall this function feels like it is playing whack-a-mole with various use cases. I feel like we just be able to use urlencode(html_entity_decode(strip_tags($string))) once we have done any special handling for particular tags.